lxd/images.go

```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
+		if len(nodeIDs) > 1 {
+			selectedNode, err := shared.GetStableRandomInt64FromList(len(images), nodeIDs)
+			if err != nil {
+				logger.Debug("Failed to get stable random value", log.Ctx{"err": err})
+				continue
+			}
+
+			if d.cluster.GetNodeID() != selectedNode {
+				continue
+			}
+		}
+
+		var newImage *api.Image
+		ch := make(chan struct{})
+		go func() {
+			var deleteIDs []int
+
+			for _, image := range images {
+				_, imageInfo, err := d.cluster.GetImage(image.Project, image.Fingerprint, image.Public)
+				if err != nil {
+					logger.Error("Failed to get image", log.Ctx{"err": err})
+				}
+
+				newInfo, err := autoUpdateImage(d, nil, image.ID, imageInfo, image.Project)
+				if err != nil {
+					logger.Error("Failed to update image", log.Ctx{"err": err})
+				} else {
+					deleteIDs = append(deleteIDs, image.ID)
+				}
+
+				if newImage == nil {
+					newImage = newInfo
+				}
+			}
+
+			if newImage != nil {
+				err := distributeImage(d, nodes, fingerprint, newImage)
+				if err != nil {
+					logger.Error("Failed to distribute image", log.Ctx{"err": err, "fingerprint": newImage.Fingerprint})
+				}
+
+				for _, ID := range deleteIDs {
+					// Remove the database entry for the image.
+					if err = d.cluster.DeleteImage(ID); err != nil {
+						logger.Debugf("Error deleting image from database: %s", err)
+					}
+				}
+			}
+
+			ch <- struct{}{}
+		}()
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ch:
 		}
 	}
 
 	return nil
 }
 
-func autoUpdateImagesInProject(ctx context.Context, d *Daemon, projectName string) error {
-	var err error
-	var interval int64
+func distributeImage(d *Daemon, nodes []string, oldFingerprint string, newImage *api.Image) error {
+	// Skip own node
+	address, _ := node.ClusterAddress(d.db)
 
-	project, err := d.cluster.GetProject(projectName)
+	// Get the IDs of all storage pools on which a storage volume
+	// for the requested image currently exists.
+	poolIDs, err := d.cluster.GetPoolsWithImage(newImage.Fingerprint)
 	if err != nil {
+		logger.Error("Error getting image pools", log.Ctx{"err": err, "fingerprint": oldFingerprint})
 		return err
 	}
 
-	if project.Config["images.auto_update_interval"] != "" {
-		interval, err = strconv.ParseInt(project.Config["images.auto_update_interval"], 10, 64)
+	// Translate the IDs to poolNames.
+	poolNames, err := d.cluster.GetPoolNamesFromIDs(poolIDs)
+	if err != nil {
+		logger.Error("Error getting image pools", log.Ctx{"err": err, "fingerprint": oldFingerprint})
+		return err
+	}
+
+	for _, nodeAddress := range nodes {
+		if nodeAddress == address {
+			continue
+		}
+
+		var nodeInfo db.NodeInfo
+		err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			var err error
+			nodeInfo, err = tx.GetNodeByAddress(nodeAddress)
+			return err
+		})
 		if err != nil {
-			return errors.Wrap(err, "Unable to fetch project configuration")
+			logger.Errorf("Failed to retrieve the information of node: %v", err)
```

Comment:
@monstermunchkin adding the node address here would be good, could use logger.Errorf("Failed to retrieve the information of cluster member %q: %v", nodeAddress)